File: events/2015/02/solution.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"time"
	"yod_aoc/utils/go/aoc_utils"
)

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
func part1(input []string) int {
	totalArea := 0

	for _, present := range input {
		sides := strings.SplitN(present, "x", -1)
		//fmt.Println(sides)
		length, lerr := strconv.Atoi(sides[0])
		width, werr := strconv.Atoi(sides[1])
		height, herr := strconv.Atoi(sides[2])

		if lerr != nil || werr != nil || herr != nil {
			continue
		}

		area := (2 * length * width) + (2 * width * height) + (2 * height * length)

		smallestSide := math.Min(float64(length*width), float64(width*height))
		smallestSide = math.Min(smallestSide, float64(height*length))

		area = area + int(smallestSide)

		totalArea += area
		//fmt.Println(present)
	}

	return totalArea
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
func part2(input []string) int {
	totalRibbon := 0

	for _, present := range input {
		sides := strings.SplitN(present, "x", -1)
		// fmt.Println(sides)
		length, lerr := strconv.Atoi(sides[0])
		width, werr := strconv.Atoi(sides[1])
		height, herr := strconv.Atoi(sides[2])

		if lerr != nil || werr != nil || herr != nil {
			continue
		}

		perimeter := 2 * int(math.Min(math.Min(float64(length+width), float64(width+height)), float64(height+length)))
		volume := length * width * height

		totalRibbon = totalRibbon + perimeter + volume
		// fmt.Println(present)
	}

	return totalRibbon
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
func main() {
	// Read the arguments
	args := os.Args[1:]
	var fileName string
	if len(args) > 0 {
		fileName = args[0]
	} else {
		fileName = "input.txt"
	}
	// Start the timer!
	startTime := time.Now()

	// Read the input to an array of strings
	path := aoc_utils.GetFilePath(2015, 2, fileName)
	content, err := aoc_utils.ReadInputAsArray(path)
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	// Run the Parts
	p1Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part1(content)
	})

	p2Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part2(content)
	})

	// End the Timer!
	endTime := time.Now()
	duration := endTime.Sub(startTime)

	resultsForDay := aoc_utils.AOCDayResults{
		Year:     2015,
		Day:      2,
		Part1:    p1Result,
		Part2:    p2Result,
		Duration: duration.Nanoseconds(),
		Timestamp: aoc_utils.AOCTimestamp{
			Start: startTime.Format(time.RFC3339),
			End:   endTime.Format(time.RFC3339),
		},
	}
	jsonResults, _ := json.Marshal(resultsForDay)
	fmt.Println(string(jsonResults))
}

File: events/2015/02/solution.rs
use aoc_utils::read_input_as_array;
use aoc_utils::measure_performance;
use aoc_utils::AOCDayResults;
use aoc_utils::AOCTimestamp;
use chrono::{DateTime, Utc};
use std::convert::TryInto;

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
fn part1(input: &Vec<String>) -> i32 {
  let mut total_area: i32 = 0;

  for present in input {
    let [length, width, height]: [i32; 3] = present
        .split('x')
        .map(|s| s.parse().unwrap())
        .collect::<Vec<i32>>()
        .try_into()
        .unwrap();

    let mut area: i32 = (2 * length * width) + (2 * width * height) + (2 * height * length);

    let smallest_side: i32 = std::cmp::min(std::cmp::min(length * width, width * height), height * length);
    area += smallest_side;

    total_area += area;
  }

  total_area
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
fn part2(input: &Vec<String>) -> i32 {
  let mut total_ribbon: i32 = 0;

  for present in input {
    let [length, width, height]: [i32; 3] = present
        .split('x')
        .map(|s| s.parse().unwrap())
        .collect::<Vec<i32>>()
        .try_into()
        .unwrap();

    let perimeter = 2 * std::cmp::min(std::cmp::min(length + width, width + height), height + length);
    let volume = length * width * height;

    total_ribbon += perimeter + volume;
  }

  total_ribbon
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
fn main() {
  // Read the arguments
  let args = std::env::args().collect::<Vec<String>>();
  let default_file_name = String::from("input.txt");
  let file_name = args.get(1).unwrap_or(&default_file_name);

  // Start the timer
  let start_timestamp = std::time::SystemTime::now();
  let start_time = std::time::Instant::now();
  
  // Read the input
  let input = read_input_as_array(file_name).unwrap();

  // Run the Parts
  let p1_result = measure_performance(part1, &input);
  let p2_result = measure_performance(part2, &input);

  // Stop the timer
  let end_time = std::time::Instant::now();
  let end_timestamp = std::time::SystemTime::now();
  let duration = end_time.duration_since(start_time).as_nanos();

  let results = AOCDayResults {
    year: 2015,
    day: 2,
    part1: p1_result,
    part2: p2_result,
    duration: duration,
    timestamp: AOCTimestamp {
      start: DateTime::<Utc>::from(start_timestamp).to_rfc3339(),
      end: DateTime::<Utc>::from(end_timestamp).to_rfc3339(),
    },
  };

  let json_results = serde_json::to_string(&results).unwrap();
  println!("{}", json_results);
}